src/frequenz/sdk/util/ringbuffer.py

```diff
@@ -0,0 +1,259 @@
+# License: MIT
+# Copyright © 2022 Frequenz Energy-as-a-Service GmbH
+
+"""Ringbuffer implementation with focus on time & memory efficiency."""
+
+from __future__ import annotations
+
+from copy import deepcopy
+from datetime import datetime, timedelta
+from typing import Any, Generic, Sequence, TypeVar, Union
+
+import numpy as np
+
+T = TypeVar("T")
+
+Container = Union[list, np.ndarray]
+
+
+class OrderedRingBuffer(Generic[T]):
+    """Time aware ringbuffer that keeps its entries sorted by time."""
+
+    def __init__(
+        self,
+        buffer: Any,
+        sample_rate: float,
+        window_border: datetime = datetime(1, 1, 1),
+    ) -> None:
+        """Initialize the time aware ringbuffer.
+
+        Args:
+            buffer: instance of a buffer container to use internally
+            sample_rate: resolution of the incoming timestamps in
+                hz (samples per second)
+            window_border: datetime depicting point in time to use as border
+                beginning, useful to make data start at the beginning of the day or
+                hour.
+        """
+        self._buffer = buffer
+        self._sample_rate = sample_rate
+        self._window_start = window_border
+
+        self._missing_windows = []
+        self._datetime_newest = datetime.min
+        self._datetime_oldest = datetime.max
+
+    @property
+    def missing_windows(self) -> list:
+        """Get the list of missing windows.
+
+        Returns:
+            list of missing windows.
+        """
+        return self._missing_windows
+
+    @property
+    def maxlen(self) -> int:
+        """Get the max length.
+
+        Returns:
+            the max amount of items this container can hold.
+        """
+        return len(self._buffer)
+
+    def update(self, timestamp: datetime, value: T, missing: bool = False) -> bool:
+        """Update the buffer with a new value for the given timestamp.
+
+        Args:
+            timestamp: Timestamp of the new value
+            value: value to add
+            missing: if true, the given timestamp will be recorded as missing.
+                The value will still be written.
+
+        Returns:
+            Nothing.
+        """
+        # Don't add outdated entries
+        if timestamp < self._datetime_oldest and self._datetime_oldest != datetime.max:
+            return False
+
+        # Update timestamps
+        prev_newest = self._datetime_newest
+        self._datetime_newest = max(self._datetime_newest, timestamp)
+        self._datetime_oldest = self._datetime_newest - timedelta(
+            seconds=(len(self._buffer) - 1) / self._sample_rate
+        )
+
+        assert self.datetime_to_index(self._datetime_newest) != self.datetime_to_index(
+            self._datetime_oldest
+        )
```

Comment:
Why is it needed?

Reply:
It's not "needed", it's just to make sure our timestamps are what we expect them to be.